### PR TITLE
NNX: save checkpoints in the Linen on-disk layout (interchangeable with Linen)

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## Summary

A `pure_nnx` run now writes the same on-disk checkpoint layout as a Linen run, so Linen and NNX checkpoints are interchangeable. After the NNX migration this lets existing Linen checkpoints keep working, and lets a checkpoint saved by either framework be loaded by the other.

The conversion lives on `TrainStateNNX` (no separate load-time adapter): the state serializes to the Linen layout on save and reads it back on load.

## Why
NNX's natural `nnx.state(...).to_pure_dict()` differs from Linen's checkpoint in three ways, none of which are in the model/decoder modules — the weight names already match. They come from the train state and optimizer:
1. top-level keys: NNX `{model, optimizer}` vs Linen `{params, step, opt_state}`
2. weights: NNX `model/...` vs Linen `params/params/...` (the Linen `params` collection)
3. `opt_state`: NNX serializes the optax chain as an int-keyed dict; Linen as a list with `None` for each `EmptyState`, and wraps `mu`/`nu` in the `params` collection

Plus `step` dtype (NNX `uint32` vs Linen `int32`) and NNX-only `rngs`/`dropout` state that Linen never had.

## What changes
Two source files plus their tests:
- `src/maxtext/layers/train_state_nnx.py` — `to_linen_checkpoint_dict` (save) and `from_linen_checkpoint_dict` (load) reshape between the two layouts, handling the three differences above, the `step` dtype, and dropping NNX-only `rngs`/`dropout`.
- `src/maxtext/common/checkpointing.py`:
  - save: `maybe_save_checkpoint` writes `to_linen_checkpoint_dict(state.to_pure_dict())`.
  - load (NNX target): all three paths reshape the Linen layout back into the NNX state and re-initialize the dropped `rngs`/`dropout` from defaults — `load_state_if_possible` (auto-resume), `_load_full_state_from_path` (`load_full_state_path`), and `load_params_from_path` (`load_parameters_path`).

Linen runs are unchanged; they already read and write this layout.

## Test plan
**Byte-exactness.** A `pure_nnx` gemma2-2b checkpoint is structurally identical to a
Linen one — same 76 leaves, no path or shape/dtype differences.

**Compatibility matrix.** For each model: save with both frameworks, then load each
of the three mechanisms in both directions (Linen→NNX and NNX→Linen). All eight
phases pass on six models on v6e-8:

| Model | default (auto-resume) | weight-only (`load_parameters_path`) | full-state (`load_full_state_path`) |
|-------|-----------------------|--------------------------------------|-------------------------------------|
| gpt3-52k  | ✓ | ✓ | ✓ |
| gpt3-6b   | ✓ | ✓ | ✓ |
| gemma2-2b | ✓ | ✓ | ✓ |
| gemma2-9b | ✓ | ✓ | ✓ |
| llama2-7b | ✓ | ✓ | ✓ |
| qwen3-8b  | ✓ | ✓ | ✓ |

(3 model families, both optimizer types: adam_pax single-state and adamw 3-element chain.)

**Unit tests.** `tests/unit/train_state_nnx_checkpoint_test.py` and `tests/unit/checkpointing_nnx_load_test.py` cover the converters (round-trip, opt_state list/dict, `step` dtype, RNG strip) and a real Orbax round-trip for the weight-only NNX load. 16 tests pass.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by bvandermoon — COMMENTED** [review]
Are we able to make model code changes so that the Linen/NNX checkpoints are the same instead?

File: src/maxtext/layers/train_state_nnx.py [hunk 1]
```diff
@@ -46,3 +48,138 @@ def apply_gradients(self, grads: Any):
...
+  if isinstance(step, jax.ShapeDtypeStruct):
+    return jax.ShapeDtypeStruct(step.shape, dtype, sharding=getattr(step, "sharding", None))
+  return jnp.asarray(step, dtype=dtype)
+
+
+def _strip_rng_state(tree):
+  """Removes the NNX-only 'rngs'/'dropout' subtrees that Linen doesn't carry.
+
+  Subtrees that become empty after stripping are dropped too, so the result has
+  no keys Linen wouldn't also have.2
```
Comment (bvandermoon) [inline src/maxtext/layers/train_state_nnx.py]: Nit: extra 2 added at the end of this sentence

**Review by bvandermoon — APPROVED** [review]

File: src/maxtext/layers/train_state_nnx.py (hunk 1 above)
Comment (ecnal-cienet) [reply src/maxtext/layers/train_state_nnx.py]: Removed. Thanks.

**Review by NuojCheng — APPROVED** [review]
